File: app/src/main/java/com/zh/touchassistant/ui/fragment/AutoHideSettingFragment.java
package com.zh.touchassistant.ui.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.BaseViewHolder;
import com.lzh.easythread.EasyThread;
import com.zh.touchassistant.R;
import com.zh.touchassistant.database.biz.IAutoHideFloatBiz;
import com.zh.touchassistant.database.biz.impl.AutoHideFloatBiz;
import com.zh.touchassistant.database.model.dto.AutoHideFloatDTO;
import com.zh.touchassistant.database.model.vo.AutoHideFloatVO;
import com.zh.touchassistant.model.AutoHideModel;
import com.zh.touchassistant.model.InstallAppInfoModel;
import com.zh.touchassistant.provider.ContextProvider;
import com.zh.touchassistant.util.AppInfoUtil;
import com.zh.touchassistant.util.singleton.ISingletonStorage;
import com.zh.touchassistant.widget.SwitchButton;

import java.util.ArrayList;
import java.util.List;

/**
 * <b>Package:</b> com.zh.touchassistant.ui.fragment <br>
 * <b>FileName:</b> AutoHideSettingFragment <br>
 * <b>Create Date:</b> 2018/12/17  下午4:19 <br>
 * <b>Author:</b> zihe <br>
 * <b>Description:</b>  <br>
 */
public class AutoHideSettingFragment extends Fragment {
    private RecyclerView mRecyclerView;
    private List<AutoHideModel> mDatas;
    private ProgressBar mWaitProgressBar;

    public static AutoHideSettingFragment newInstance() {
        return new AutoHideSettingFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_auto_hide, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ISingletonStorage storage = (ISingletonStorage) ContextProvider.get().getApplication();
        final AutoHideFloatBiz biz = storage.getInstance(IAutoHideFloatBiz.class, AutoHideFloatBiz.class);
        mRecyclerView = view.findViewById(R.id.recycler_view);
        mWaitProgressBar = view.findViewById(R.id.wait_progress_bar);
        mDatas = new ArrayList<>();
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        final BaseQuickAdapter adapter = new BaseQuickAdapter<AutoHideModel, BaseViewHolder>(R.layout.item_local_install_app, mDatas) {
            @Override
            protected void convert(BaseViewHolder helper, final AutoHideModel item) {
                helper.setImageDrawable(R.id.app_icon_iv, item.getAppIcon());
                helper.setText(R.id.app_name_tv, item.getAppName());
                SwitchButton switchButton = helper.getView(R.id.auto_hide_switch);
                switchButton.setChecked(item.isAutoHide());
                if (switchButton.getCheckedChangeListener() == null) {
                    switchButton.setOnCheckedChangeListener(new SwitchButton.OnCheckedChangeListener() {
                        @Override
                        public void onCheckedChanged(SwitchButton button, boolean isChecked) {
                            //取消自动隐藏
                            AutoHideFloatDTO dto = new AutoHideFloatDTO();
                            dto.setAppPackageName(item.getPackageName());
                            if (isChecked) {
                                biz.addAutoHide(dto);
                            } else {
                                biz.removeAutoHide(dto);
                            }
                        }
                    });
                }
            }
        };
        mRecyclerView.setAdapter(adapter);
        mRecyclerView.addItemDecoration(new DividerItemDecoration(getContext(), RecyclerView.VERTICAL));
        mWaitProgressBar.setVisibility(View.VISIBLE);
        EasyThread.Builder.createCacheable().build().execute(new Runnable() {
            @Override
            public void run() {
                //本机App列表
                List<InstallAppInfoModel> installAppInfoList = AppInfoUtil
                        .getInstallAppInfoList(getContext(), true);
                List<String> autoHideAppPackageNameList = getAutoHideAppPackageNameList();
                //过滤掉本应用
                autoHideAppPackageNameList.remove(getContext().getPackageName());
                for (InstallAppInfoModel installAppInfoModel : installAppInfoList) {
                    AutoHideModel autoHideModel = new AutoHideModel();
                    //是自动隐藏
                    if (autoHideAppPackageNameList.contains(installAppInfoModel.getPackageName())) {
                        autoHideModel.setAutoHide(true);
                    } else {
                        autoHideModel.setAutoHide(false);
                    }
                    autoHideModel.setAppIcon(installAppInfoModel.getAppIcon());
                    autoHideModel.setAppName(installAppInfoModel.getAppName());
                    autoHideModel.setPackageName(installAppInfoModel.getPackageName());
                    mDatas.add(autoHideModel);
                }
                adapter.notifyDataSetChanged();
                FragmentActivity activity = getActivity();
                if (activity != null && !activity.isFinishing()) {
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            mWaitProgressBar.setVisibility(View.GONE);
                        }
                    });
                }
            }
        });
    }

    private List<String> getAutoHideAppPackageNameList() {
        ArrayList<String> result = new ArrayList<>();
        //已经存在的数据
        ISingletonStorage storage = (ISingletonStorage) ContextProvider.get().getApplication();
        AutoHideFloatBiz biz = storage.getInstance(IAutoHideFloatBiz.class, AutoHideFloatBiz.class);
        List<AutoHideFloatVO> autoHideAppList = biz.getAutoHideAppList();
        for (AutoHideFloatVO vo : autoHideAppList) {
            result.add(vo.getAppPackageName());
        }
        return result;
    }
}

File: app/src/main/java/com/zh/touchassistant/ui/fragment/AutoHideSettingFragment.kt
package com.zh.touchassistant.ui.fragment

import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import com.lzh.easythread.EasyThread
import com.pnikosis.materialishprogress.ProgressWheel
import com.zh.touchassistant.R
import com.zh.touchassistant.base.BaseTouchAssistantFragment
import com.zh.touchassistant.database.biz.IAutoHideFloatBiz
import com.zh.touchassistant.database.biz.impl.AutoHideFloatBiz
import com.zh.touchassistant.ext.isNull
import com.zh.touchassistant.ext.setGone
import com.zh.touchassistant.ext.setVisible
import com.zh.touchassistant.itembinder.AutoHideViewBinder
import com.zh.touchassistant.listener.DelayOnClickListener
import com.zh.touchassistant.model.AutoHideModel
import com.zh.touchassistant.provider.ContextProvider
import com.zh.touchassistant.util.AppInfoUtil
import com.zh.touchassistant.util.singleton.ISingletonStorage
import com.zh.touchassistant.widget.TopBar
import kotterknife.bindView
import me.drakeet.multitype.MultiTypeAdapter

/**
 * **Package:** com.zh.touchassistant.ui.fragment <br></br>
 * **FileName:** AutoHideSettingFragment <br></br>
 * **Create Date:** 2018/12/17  下午4:19 <br></br>
 * **Author:** zihe <br></br>
 * **Description:**  <br></br>
 */
class AutoHideSettingFragment : BaseTouchAssistantFragment() {
    private val mTopBar: TopBar by bindView(R.id.top_bar)
    private val mRecyclerView: RecyclerView by bindView(R.id.recycler_view)
    private val mWaitProgressWheel: ProgressWheel by bindView(R.id.wait_progress_wheel)
    private val mAdapter: MultiTypeAdapter by lazy {
        val typeAdapter = MultiTypeAdapter(mDatas)
        typeAdapter.register(AutoHideModel::class.java, AutoHideViewBinder())
        return@lazy typeAdapter
    }
    private val mDatas: MutableList<AutoHideModel> by lazy {
        mutableListOf<AutoHideModel>()
    }

    //之前设置的数据
    private val autoHideAppPackageNameList: MutableList<String>
        get() {
            val result = mutableListOf<String>()
            val storage = ContextProvider.get().application as ISingletonStorage
            val biz = storage.getInstance<AutoHideFloatBiz>(IAutoHideFloatBiz::class.java, AutoHideFloatBiz::class.java)
            val autoHideAppList = biz.autoHideAppList
            for (vo in autoHideAppList) {
                result.add(vo.appPackageName)
            }
            return result
        }

    override fun onLayoutId(): Int {
        return R.layout.fragment_auto_hide
    }

    override fun onLayoutAfter() {
        super.onLayoutAfter()
        if (context.isNull()) return
        val context = context!!
        mTopBar.apply {
            setTitle("隐藏悬浮球")
            addLeftBackImageButton().setOnClickListener(object : DelayOnClickListener() {
                override fun onDelayClick(view: View) {
                    activity?.finish()
                }
            })
        }
        mRecyclerView.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = mAdapter
            addItemDecoration(DividerItemDecoration(context, RecyclerView.VERTICAL))
        }
        mWaitProgressWheel.setVisible()
        EasyThread.Builder.createCacheable().build().execute {
            //本机App列表
            val installAppInfoList = AppInfoUtil
                    .getInstallAppInfoList(context, true)
            val autoHideAppPackageNameList = autoHideAppPackageNameList
            //过滤掉本应用
            autoHideAppPackageNameList.remove(context.packageName)
            for (installAppInfoModel in installAppInfoList) {
                val autoHideModel = AutoHideModel()
                //是自动隐藏
                autoHideModel.isAutoHide = autoHideAppPackageNameList.contains(installAppInfoModel.packageName)
                autoHideModel.appIcon = installAppInfoModel.appIcon
                autoHideModel.appName = installAppInfoModel.appName
                autoHideModel.packageName = installAppInfoModel.packageName
                mDatas.add(autoHideModel)
            }
            val activity = activity
            if (activity != null && !activity.isFinishing) {
                activity.runOnUiThread {
                    mAdapter.notifyDataSetChanged()
                    mWaitProgressWheel.setGone()
                }
            }
        }
    }

    companion object {
        fun newInstance(): AutoHideSettingFragment {
            return AutoHideSettingFragment()
        }
    }
}